### PR TITLE
Python Wrapper: Reference tests

### LINKED ISSUE #6991: Python Wrapper: Implement Reference System Tests
Add the missing following coverage once Branch and Object are implemented:

```
def test_reference_log():
    pass
    # TODO: Implement


def test_reference_diff():
    pass
    # TODO: Implement


def test_reference_merge_into():
    pass
    # TODO: Implement
```

### LINKED ISSUE #7004: Python Wrapper: Should we remove the max_amount param from generators?
Pros:
- Simplifying user interface

Cons:
- Might cause confusion and create a strange behavior when providing additional listing args

From discussions:
https://github.com/treeverse/lakeFS/pull/6992#discussion_r1395340315

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #7004
Closes #6991

### Background

Add reference integration tests
Bug fixes and modifications as result of #7004

### CODE REVIEW

File: clients/python-wrapper/lakefs/branch.py [hunk 1]
```diff
@@ -18,7 +18,16 @@ class Branch(Reference):
...
     """
 
-    def create(self, source_reference_id: str, exist_ok: bool = False) -> Branch:
+    def _get_commit(self):
+        """
+        For branches override the default _get_commit method to ensure we always fetch the latest head
+        """
+        with api_exception_handler():
+            commit = self._client.sdk_client.commits_api.get_commit(self._repo_id, self._id)
+            self._commit = Commit(**commit.__dict__)
```
Comment (ozkatz) [inline clients/python-wrapper/lakefs/branch.py:27]: Why are we caching the current HEAD? The point of a branch is to allow mutation and for HEAD to change.. Perhaps I'm missing something.

Comment (ozkatz) [reply clients/python-wrapper/lakefs/branch.py:27]:
```suggestion
            self._commit = Commit(**commit.dict())
```

Pydantic gives you a public method for this

Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py:27]: Prefer to defer it to another PR where I'll change all occurrences of this code fragment

Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py:27]: Done

Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py:27]: > Why are we caching the current HEAD? The point of a branch is to allow mutation and for HEAD to change.. Perhaps I'm missing something.

We're not caching the branch head. the self._commit is inherited from the Reference class. For commits and tags which are immutable we don't to fetch the commit information each time so we do it only once. In branch we want to override this behavior by explicitly getting the head each time

Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py:27]: Cleaned up the code

**Pull request merged**